geo/src/algorithm/affine_ops.rs

```diff
@@ -111,16 +115,16 @@ impl<T: CoordNum, M: MapCoordsInPlace<T> + MapCoords<T, T, Output = Self>> Affin
 /// ], max_relative = 1.0);
 /// ```
 #[derive(Copy, Clone, PartialEq, Eq)]
-pub struct AffineTransform<T: CoordNum = f64>([[T; 3]; 3]);
+pub struct AffineTransform<T: CoordNum + Neg = f64>([[T; 3]; 3]);
```

Comment:
Do we actually need to add these `Neg` bounds?

Reply:
In some form, because they're needed by the arithmetic that produces the determinant (according to the compiler).
But I haven't tried @michaelkirk's suggestion (see his previous comment) yet.

Reply:
But it's only needed for `inverse`, right? It still seems to build if I split out `inverse` into its own `impl` block and drop the new `Neg` bounds from everywhere else:

```rust
impl<T: CoordNum + Neg> AffineTransform<T> {
    /// Return the inverse of a given transform. Composing a transform with its inverse yields
    /// the [identity matrix](Self::identity)
    pub fn inverse(&self) -> Option<Self>
    where
        <T as Neg>::Output: Mul<T>,
        <<T as Neg>::Output as Mul<T>>::Output: ToPrimitive,
    {
        let a = self.0[0][0];
        let b = self.0[0][1];
        let xoff = self.0[0][2];
        let d = self.0[1][0];
        let e = self.0[1][1];
        let yoff = self.0[1][2];

        let determinant = a * e - b * d;

        if determinant == T::zero() {
            return None; // The matrix is not invertible
        }

        let inv_det = T::one() / determinant;
        Some(Self::new(
            e * inv_det,
            T::from(-b * inv_det).unwrap(),
            (b * yoff - e * xoff) * inv_det,
            T::from(-d * inv_det).unwrap(),
            a * inv_det,
            (d * xoff - a * yoff) * inv_det,
        ))
    }
}
```

PS: should probably be `#[must_use]` too.

Reply:
Both of these are good points!

Reply:
(Though I'd recommend doing 

```
impl<T: CoordNum + Neg<Output=T>> AffineTransform<T> {
```

Reply:
> But it's only needed for `inverse`, right? It still seems to build if I split out `inverse` into its own `impl` block and drop the new `Neg` bounds from everywhere else:
> 
> 
> 
> ```rust
> 
> impl<T: CoordNum + Neg> AffineTransform<T> {
> 
>     /// Return the inverse of a given transform. Composing a transform with its inverse yields
> 
>     /// the [identity matrix](Self::identity)
> 
>     pub fn inverse(&self) -> Option<Self>
> 
>     where
> 
>         <T as Neg>::Output: Mul<T>,
> 
>         <<T as Neg>::Output as Mul<T>>::Output: ToPrimitive,
> 
>     {
> 
>         let a = self.0[0][0];
> 
>         let b = self.0[0][1];
> 
>         let xoff = self.0[0][2];
> 
>         let d = self.0[1][0];
> 
>         let e = self.0[1][1];
> 
>         let yoff = self.0[1][2];
> 
> 
> 
>         let determinant = a * e - b * d;
> 
> 
> 
>         if determinant == T::zero() {
> 
>             return None; // The matrix is not invertible
> 
>         }
> 
> 
> 
>         let inv_det = T::one() / determinant;
> 
>         Some(Self::new(
> 
>             e * inv_det,
> 
>             T::from(-b * inv_det).unwrap(),
> 
>             (b * yoff - e * xoff) * inv_det,
> 
>             T::from(-d * inv_det).unwrap(),
> 
>             a * inv_det,
> 
>             (d * xoff - a * yoff) * inv_det,
> 
>         ))
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> PS: should probably be `#[must_use]` too.

Ahh, it didn't occur to me to split it into its own impl. This is much better I think.

Reply:
We could also return `None` if the cast fails..

Reply:
Oh and breaking out this impl means it's no longer a breaking change, nice!